### ISSUE TITLE
dataclasses init argument order depends on slot order

### ISSUE THREAD

**Issue by Gobot1234** [microsoft/pyright#3440]
**Describe the bug**
The order of a dataclass's init depends on the order of slots if they are defined

**To Reproduce**
```py
import dataclasses


@dataclasses.dataclass
class Foo:
    __slots__ = ("y", "x")  # removing this makes it work
    x: int
    y: str


print(Foo(1, "bar"))
```
```
11:11  error   Argument of type "Literal[1]" cannot be assigned to parameter "y" of type "str" in function "__init__" ​Pyright:
                    "Literal[1]" is incompatible with "str"
11:14  error   Argument of type "Literal['bar']" cannot be assigned to parameter "x" of type "int" in function "__init__" ​Pyright:
                    "Literal['bar']" is incompatible with "int"
```

**Expected behavior**
No error. Running this code produces
```py
Foo(x=1, y='bar')
```
which does not depend on the order of the slots.

**VS Code extension or command-line**
1.1.245

**Comment by erictraut** [comment]
I'm not sure this edge case is worth addressing. Slots are already used rarely with dataclasses, and they're used only by very experienced developers. The obvious workaround in this case is to reorder the list of items in the `__slots__` to match the ordering of your field declarations. Does that work for you?

**Comment by Gobot1234** [comment]
I don't particularly mind changing them no, but I'm naively not entirely sure how this is an edge case and why the order depends on the presence of slots.

**Comment by erictraut** [comment]
By "edge case", I mean that this is probably a case that only you and maybe one or two other developers will ever encounter. I'll take a look at addressing this, but if it requires adding significant special-case logic, I don't think it's worth it.

**Comment by erictraut** [comment]
I found a straightforward way to address this. The fix will be included in the next release.

**Comment by JelleZijlstra** [comment]
I'm not sure this is intentional behavior on CPython's side. Does it work this way across all Python versions?

I don't see any unit tests for this behavior in CPython. It may be worth adding some so we ensure this behavior remains stable in the future.

**Comment by Gobot1234** [comment]
> I'm not sure this is intentional behavior on CPython's side. Does it work this way across all Python versions?

I ran this snippet on both 3.7.4 and 3.10.4 and they both produced the same result. I could ask Eric Smith if you wanted.

**Comment by JelleZijlstra** [comment]
Thanks, then I think we should add a unit test that tests this behavior.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.246, which I just published. It will also be included in the next release of pylance.

**Comment by Gobot1234** [comment]
Thank you
